python/lsst/pipe/tasks/make_psf_matched_warp.py

```diff
@@ -0,0 +1,215 @@
+    def matchingKernelSize(self, value):
+        self.psfMatch.kernel.active.kernelSize = value
+
+    def setDefaults(self):
+        super().setDefaults()
+        self.psfMatch.kernel["AL"].alardSigGauss = [1.0, 2.0, 4.5]
+        self.modelPsf.defaultFwhm = 7.7
+
+
+class MakePsfMatchedWarpTask(PipelineTask):
+    ConfigClass = MakePsfMatchedWarpConfig
+    _DefaultName = "makePsfMatchedWarp"
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.makeSubtask("psfMatch")
+
+    @timeMethod
+    def run(self, direct_warp: Exposure):
+        """Make a PSF-matched warp from a direct warp.
+
+        Each individual detector from the direct warp is isolated, one at a
+        time, and PSF-matched to the same model PSF. The PSF-matched images are
+        then added back together to form the final PSF-matched warp. The bulk
+        of the work is done by the `psfMatchTask`.
+
+        Notes
+        -----
+        Pixels that receive no inputs are set to NaN, for e.g, chip gaps. This
+        violates LSST algorithms group policy, but will be fixed during the
+        coaddition stage.
+
+        Parameters
+        ----------
+        direct_warp : `lsst.afw.image.Exposure`
+            Direct warp to be PSF-matched.
+
+        Returns
+        -------
+        struct : `lsst.pipe.base.Struct`
+            Struct containing the PSF-matched warp under the attribute
+            `psf_matched_warp`.
+        """
+        modelPsf = self.config.modelPsf.apply()
+
+        # Prepare the output exposure. We clone the input image to keep the
+        # metadata, but reset the image and variance plans.
+        exposure_psf_matched = direct_warp.clone()
+        exposure_psf_matched.image.array[:, :] = np.nan
+        exposure_psf_matched.variance.array[:, :] = np.inf
+        exposure_psf_matched.setPsf(modelPsf)
+
+        bit_mask = direct_warp.mask.getPlaneBitMask("NO_DATA")
+        total_good_pixels = 0  # Total number of pixels copied to output.
+
+        for row in direct_warp.info.getCoaddInputs().ccds:
+            transform = makeWcsPairTransform(row.wcs, direct_warp.wcs)
+            warp_psf = WarpedPsf(row.getPsf(), transform)
+
+            if (destination_polygon := row.validPolygon) is None:
+                # Calculate the polygon for this detector.
+                src_polygon = Polygon(
+                    [geom.Point2D(corner) for corner in row.getBBox().getCorners()]
+                )
+                destination_polygon = src_polygon.transform(transform).intersectionSingle(
+                    geom.Box2D(direct_warp.getBBox())
+                )
+                self.log.trace("Polygon for detector=%d is calculated as %s",
+                               row["ccd"],
+                               destination_polygon
+                               )
+            else:
+                self.log.trace("Polygon for detector=%d is read from the input calexp as %s",
```

Comment:
I'd just make these logs `debug` instead of `trace`.  There wouldn't been all that many of them if they were on, and `trace` level is a little obscure.